text/3371-cargo-target-base-dir.md

```diff
@@ -0,0 +1,383 @@
+
+This naming scheme is **not considered stable**: the method will probably not change often but `cargo` offers no guarantee and may change it in any release. Tools that needs to interact with `cargo`'s target directory should not rely on its value for more than a single invocation of them: they should instead query `cargo metadata` for the actual value each time they are invoked.
+
+The path used for the naming of the final target directory is the one found *after* symlink resolution: `bazel` does it too and I have not found any complaints about this and it has the distinct advantage of allowing to make a symlink to a project somewhere else on the machine (for example to organise work projects) and avoid duplicating the build directory and all its data (which can be quite heavy).
+
+To prevent collisions by craftings paths, the `<project-path>` directory will be computed from a hash of the workspace manifest's full path (and possibly other data, for example `bazel` uses its version and the current user too).
+
+### Symbolic links
+
+In the following situation
+
+```
+/Users/
+├─ poliorcetics/
+│  ├─ projects/
+│  │  ├─ actual-crate/
+│  │  │  ├─ Cargo.toml
+│  │  ├─ symlink-to-crate/ -> actual-crate/
+```
+
+When calling `cargo metadata` in the `symlink-to-crate` path, the result contains `"manifest_path": "/Users/poliorcetics/projects/actual-crate/Cargo.toml"` and `"workspace_root":"/Users/poliorcetics/projects/actual-crate"`. This behaviour means that symlinks won't change the final directory used inside `CARGO_TARGET_BASE_DIR`, or in other words: symbolic links are resolved.
+
+### Handle possibly thousands of directories in a single `CARGO_TARGET_BASE_DIR` path
+
+While a single dev machine is unlikely to have enough projects that the scheme of `<project-path>` will produce enough directories to slow down working in `$CARGO_TARGET_BASE_DIR/`, it could still happen, and notably in private CI, which are often less compartimentalized than public ones. Simple cruft over time (i.e, never calling `cargo clean` over years) could also make it happen, if much slower.
+
+A proposed solution for `cargo` to split the hash into something like `$CARG_TARGET_BASE_DIR/hash[:2]/hash[2:4]/hash[4:]/...`. Since the naming scheme is considered an implementation detail, if this prove insufficient it could be change in a subsequent version of `cargo`.
+
+## Impact on `cargo ...` calls
+
+When calling `cargo` where `CARGO_TARGET_BASE_DIR` is active, `CARGO_TARGET_DIR` is set by all `cargo` calls that happen in a Cargo workspace, as much as possible. For third party tools (`cargo-*`), where cargo does not know about the relevant `Cargo.toml`, the tool would have to use [`cargo_metadata`](https://docs.rs/cargo_metadata), as is already expected today.
```

Comment:
I'm still not quite clear what it means exactly. Is it setting the `CARGO_TARGET_DIR` environment variable for build scripts? Proc macros? rustc? If so, why would it do that? None of those should be reading the CARGO_TARGET_DIR environment variable. When calling external `cargo-*` extensions, cargo doesn't know anything about manifests so it wouldn't be able to set the correct path. 

I'm not quite following the example for `cargo install` since it doesn't seem to be referring to an environment variable.

Reply:
Fixed again